libs/pdi-scanner/src/main.rs

```diff
@@ -0,0 +1,349 @@
+        .with(
+            tracing_subscriber::EnvFilter::builder()
+                .with_default_directive(
+                    format!(
+                        "{}={}",
+                        env!("CARGO_BIN_NAME").replace('-', "_"),
+                        config.log_level
+                    )
+                    .parse()?,
+                )
+                .from_env_lossy(),
+        )
+        .with(stderr_log)
+        .init();
+
+    Ok(())
+}
+
+#[derive(Debug, serde::Deserialize)]
+#[serde(tag = "type")]
+#[serde(rename_all = "camelCase")]
+enum Command {
+    #[serde(rename = "exit")]
+    Exit,
+
+    #[serde(rename = "connect")]
+    Connect,
+
+    #[serde(rename = "disconnect")]
+    Disconnect,
+
+    #[serde(rename = "get_scanner_status")]
+    GetScannerStatus,
+
+    #[serde(rename = "enable_scanning")]
+    EnableScanning,
+
+    #[serde(rename = "eject")]
+    #[serde(rename_all = "camelCase")]
+    EjectDocument { eject_motion: EjectMotion },
+
+    #[serde(rename = "enable_msd")]
+    EnableMsd { enable: bool },
+
+    #[serde(rename = "calibrate_msd")]
+    #[serde(rename_all = "camelCase")]
+    CalibrateMsd {
+        calibration_type: DoubleFeedDetectionCalibrationType,
+    },
+
+    #[serde(rename = "get_msd_calibration_config")]
+    GetMsdCalibrationConfig,
+}
+
+#[derive(Debug, serde::Serialize)]
+#[serde(tag = "type")]
+enum Response {
+    #[serde(rename = "ok")]
+    Ok,
+
+    #[serde(rename = "error")]
+    Err { message: String },
+
+    #[serde(rename = "scan_complete")]
+    #[serde(rename_all = "camelCase")]
+    ScanComplete { image_data: (String, String) },
+
+    #[serde(rename = "msd_calibration_config")]
+    #[serde(rename_all = "camelCase")]
+    MsdCalibrationConfig {
+        led_intensity: u16,
+        single_sheet_calibration_value: u16,
+        double_sheet_calibration_value: u16,
+        threshold_value: u16,
+    },
+
+    #[serde(rename = "scanner_status")]
+    #[serde(rename_all = "camelCase")]
+    ScannerStatus { status: Status },
+}
+
+fn send_response(response: &Response) -> color_eyre::Result<()> {
+    serde_json::to_writer(io::stdout(), response)?;
+    let mut stdout = io::stdout().lock();
+    stdout.write_all(b"\n")?;
+    Ok(())
+}
+
+fn main() -> color_eyre::Result<()> {
+    let config = Config::parse();
+    setup(&config)?;
+
+    let (stdin_tx, stdin_rx) = std::sync::mpsc::channel();
+
+    std::thread::spawn(move || {
+        loop {
+            let mut buffer = String::new();
+            match io::stdin().read_line(&mut buffer) {
+                Ok(_) => {
+                    let command = serde_json::from_str::<Command>(&buffer).unwrap();
+                    stdin_tx.send(command).unwrap();
+                    // println!("received message: {command:?}");
+                }
+
+                Err(e) => {
+                    tracing::error!("failed to read from stdin: {e:?}");
+                }
+            }
+        }
+    });
```

Comment:
What's the advantage to having a separate thread for checking stdio for new commands vs just doing it in the main loop?

Reply:
Reading from stdin is a synchronous operation and does not have "read with timeout" semantics, so the loop below would be unable to process events coming from the scanner until something came in over stdin to wake it up.